src/DynamoCore/DSEngine/AstBuilder.cs

```diff
@@ -83,6 +93,80 @@ public static IEnumerable<NodeModel> TopologicalSort(IEnumerable<NodeModel> node
             return sortedNodes.Where(nodeModels.Contains);
         }
 
+        /// <summary>
+        /// BFS and post-order traverse to ensure all inputs are visited in
+        /// order and before downstream node.
+        /// </summary>
+        /// <param name="node"></param>
+        /// <param name="nodeFlags"></param>
+        /// <param name="sortedNodes"></param>
```

Comment:
I'm typing these comments as a way to *teach myself* about the code, so ignore them if you want:

```csharp
/// <summary>
/// Call this method to perform BFS and post-order traverse given a node. This 
/// ensures all inputs are visited in order and before downstream node is visited.
/// </summary>
/// <param name="node">The current node in traversal.</param>
/// <param name="nodeFlags">A dictionary of nodes with their corresponding visited
/// flag. A node found in this dictionary can either be MarkFlag.TempMark (currently
/// being visited), or MarkFlag.Marked (having been visited). If the current node is 
/// not found in the dictionary, then it has not been visited before.</param>
/// <param name="sortedNodes">A list of nodes sorted from upstream nodes to 
/// downstream nodes.</param>
```

And, by the way, I'm sure we follow the naming convention strictly, the method name should have been `BfsTraverse` instead ;)